runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/stableinput/BufferingDoFnRunner.java

```diff
@@ -0,0 +1,213 @@
+          } else {
+            return NonKeyedBufferingElementsHandler.create(
+                operatorStateBackend.getListState(stateDescriptor));
+          }
+        };
+    this.currentStateId = generateNewId();
+    this.currentBufferingElementsHandler = bufferingElementsHandlerFactory.create(currentStateId);
+  }
+
+  @Override
+  public void startBundle() {
+    // Do not start a bundle, start it later when emitting elements
+  }
+
+  @Override
+  public void processElement(WindowedValue<InputT> elem) {
+    currentBufferingElementsHandler.buffer(new BufferedElements.Element(elem));
+  }
+
+  @Override
+  public void onTimer(
+      String timerId, BoundedWindow window, Instant timestamp, TimeDomain timeDomain) {
+    currentBufferingElementsHandler.buffer(
+        new BufferedElements.Timer(timerId, window, timestamp, timeDomain));
+  }
+
+  @Override
+  public void finishBundle() {
+    // Do not finish a bundle, finish it later when emitting elements
+  }
+
+  @Override
+  public DoFn<InputT, OutputT> getFn() {
+    return underlying.getFn();
+  }
+
+  /** Should be called when a checkpoint is created. */
+  public void checkpoint(long checkpointId) throws Exception {
+    // We are about to get checkpointed. The elements buffered thus far
+    // have to be added to the global CheckpointElement state which will
+    // be used to emit elements later when this checkpoint is acknowledged.
+    addToBeAcknowledgedCheckpoint(checkpointId, currentStateId);
+    currentStateId = generateNewId();
+    currentBufferingElementsHandler = bufferingElementsHandlerFactory.create(currentStateId);
+  }
+
+  /** Should be called when a checkpoint is completed. */
+  public void checkpointCompleted(long checkpointId) throws Exception {
+    List<CheckpointElement> toAck = removeToBeAcknowledgedCheckpoints(checkpointId);
+    for (CheckpointElement toBeAcked : toAck) {
+      BufferingElementsHandler bufferingElementsHandler =
+          bufferingElementsHandlerFactory.create(toBeAcked.internalId);
+      Iterator<BufferedElement> iterator = bufferingElementsHandler.getElements().iterator();
+      boolean hasElements = iterator.hasNext();
+      if (hasElements) {
+        underlying.startBundle();
+      }
+      while (iterator.hasNext()) {
+        BufferedElement bufferedElement = iterator.next();
+        bufferedElement.processWith(underlying);
+      }
+      if (hasElements) {
+        underlying.finishBundle();
+      }
+      bufferingElementsHandler.clear();
+    }
+  }
+
+  private void addToBeAcknowledgedCheckpoint(long checkpointId, String internalId)
+      throws Exception {
+    List<CheckpointElement> checkpoints = new ArrayList<>();
+    for (CheckpointElement element : notYetAcknowledgedSnapshots.get()) {
+      checkpoints.add(element);
+    }
+    checkpoints.add(new CheckpointElement(internalId, checkpointId));
+    notYetAcknowledgedSnapshots.update(checkpoints);
```

Comment:
Would this work: `notYetAcknowledgedSnapshots.addAll(singletonList(new CheckpointElement(internalId, checkpointId))`